### PR TITLE
[misc] ActionRecorder (Stage I)

### PR DESCRIPTION
**Pull request by yuanming-hu**
This PR is intentionally undocumented.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/backends/metal/kernel_manager.cpp [hunk 1]
```diff
@@ -33,6 +35,14 @@ namespace shaders {
 using KernelTaskType = OffloadedStmt::TaskType;
 using BufferEnum = KernelAttributes::Buffers;
 
+inline bool is_primary_kernel(const std::string_view &name) {
+  return name.find("jit_evaluator") == std::string::npos;
```
Comment (archibate) [inline taichi/backends/metal/kernel_manager.cpp:39]: Why not use `kern->is_evaluator`?

File: taichi/backends/metal/kernel_manager.cpp [hunk 2]
```diff
@@ -10,7 +10,9 @@
 #include "taichi/backends/metal/constants.h"
 #include "taichi/inc/constants.h"
 #include "taichi/math/arithmetic.h"
+#include "taichi/util/action_recorder.h"
 #include "taichi/python/print_buffer.h"
+#include "taichi/util/file_sequence_writer.h"
```
Comment (archibate) [inline taichi/backends/metal/kernel_manager.cpp:15]: So this only works in metal?

File: taichi/util/action_recorder.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+void ActionRecorder::start_recording_(const std::string &fn) {
+  ofs.open(fn);
+  get_instance().running = true;
+}
+
+void ActionRecorder::stop_recording() {
+  TI_ASSERT(get_instance().running);
+  get_instance().running = false;
```
Comment (k-ye) [inline taichi/util/action_recorder.cpp:36]: close ofs?

File: taichi/backends/metal/kernel_manager.cpp [hunk 4]
```diff
@@ -306,6 +345,12 @@ class HostMetalCtxBlitter {
       const auto &arg = ctx_attribs_->args()[i];
       const auto dt = arg.dt;
       char *device_ptr = base + arg.offset_in_mem;
+      if (is_primary_kernel(kernel_name_)) {
+        ActionRecorder::record(
+            "allocate_context_buffer",
```
Comment (k-ye) [inline taichi/backends/metal/kernel_manager.cpp:350]:
```suggestion
            "context_device_to_host",
```

File: taichi/util/action_recorder.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  static void start_recording(const std::string &fn);
+
+  static void stop_recording();
+
+  static bool is_recording();
+
+ private:
+  ActionRecorder();
+
+  void record_(const std::string &content,
```
[the remark this answers is not in the document]
Comment (yuanming-hu) [reply taichi/util/action_recorder.h]: Sounds good. I have change it to `ActionRecorder::get_instance().XXX` :-)

**Pull request merged**